File: Leetcode/Python/402. Remove K Digits.py
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        if num is None or num == "" or len(num) <= k:
            return "0"

        stack = []
        for i, c in enumerate(num):
            n = ord(c) - ord('0')
            while stack:
                if stack[-1] > n and len(num) - i + len(stack) > len(num) - k:
                    stack.pop()
                else:
                    break
            stack.append(n)

        result = ''.join(str(c) for c in stack)
        idx = 0
        for c in result:
            if c != '0':
                break
            idx += 1

        return "0" if idx == len(result) else result[idx:min(len(num), len(num) - k)]

# pythonic version
class Solution(object):
    def removeKdigits(self, num, k):
        """
        :type num: str
        :type k: int
        :rtype: str
        """
        if num is None or num == "" or len(num) <= k:
            return "0"

        stack = []
        for n in num:
            while stack and k and stack[-1] > n:
                stack.pop()
                k -= 1
            stack.append(n)
        print(-k)
        return ''.join(stack[:-k or None]).lstrip('0') or "0"

File: Leetcode/Java/402. Remove K Digits.java
// Given a non-negative integer num represented as a string, remove k digits from the number so that the new number is the smallest possible.
//
// Note:
// The length of num is less than 10002 and will be ≥ k.
// The given num does not contain any leading zero.
// Example 1:
//
// Input: num = "1432219", k = 3
// Output: "1219"
// Explanation: Remove the three digits 4, 3, and 2 to form the new number 1219 which is the smallest.
// Example 2:
//
// Input: num = "10200", k = 1
// Output: "200"
// Explanation: Remove the leading 1 and the number is 200. Note that the output must not contain leading zeroes.
// Example 3:
//
// Input: num = "10", k = 2
// Output: "0"
// Explanation: Remove all the digits from the number and it is left with nothing which is 0.

public class Solution {
    public String removeKdigits(String num, int k) {
        if (num == null || num.length() == 0 || k >= num.length()) {
            return "0";
        }

        int len = num.length();
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < len; i++) {
            int n = num.charAt(i) - '0';
            while (!stack.isEmpty()) {
                if (stack.peek() > n && len - i + stack.size() > len - k) {
                    stack.pop();
                } else {
                    break;
                }
            }

            stack.push(n);
        }

        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()) {
            int n = stack.pop();
            sb.append(n);
        }

        sb = sb.reverse();
        int idx = 0;
        while(idx < sb.length() && sb.charAt(idx) == '0') {
            idx++;
        }

        return idx == sb.length() ? "0" : sb.substring(idx, Math.min(len, len - k));
    }
}
